exercises/paasio/tests/paasio.rs

```diff
@@ -0,0 +1,189 @@
+extern crate paasio;
+
+/// test a few read scenarios
+macro_rules! test_read {
+    ($(#[$attr:meta])* $modname:ident ($input:expr, $len:expr)) => {
+        mod $modname {
+            use std::io::{Read, BufReader};
+            use paasio::*;
+
+            const CHUNK_SIZE: usize = 2;
+
+            $(#[$attr])*
+            #[test]
+            fn test_read_passthrough() {
+                let data = $input;
+                let size = $len(&data);
+                let mut reader = ReadStats::new(data);
+
+                let mut buffer = Vec::with_capacity(size);
+                let qty_read = reader.read_to_end(&mut buffer);
+
+                assert!(qty_read.is_ok());
+                assert_eq!(size, qty_read.unwrap() );
+                assert_eq!(size, buffer.len());
+                // 2: first to read all the data, second to check that
+                // there wasn't any more pending data which simply didn't
+                // fit into the existing buffer
+                assert_eq!(2, reader.get_reads());
+                assert_eq!(size, reader.get_bytes());
+            }
+
+            $(#[$attr])*
+            #[test]
+            fn test_read_chunks() {
+                let data = $input;
+                let size = $len(&data);
+                let mut reader = ReadStats::new(data);
+
+                let mut buffer = [0_u8; CHUNK_SIZE];
+                let mut chunks_read = 0;
+                while reader.read(&mut buffer[..]).expect("read must succeed") > 0 {
+                    chunks_read += 1;
+                }
+
+                assert_eq!(size / CHUNK_SIZE, chunks_read);
+                // we read once more than the number of chunks, because the final
+                // read returns 0 new bytes
+                assert_eq!(1+chunks_read, reader.get_reads());
+                assert_eq!(size, reader.get_bytes());
+            }
+
+            $(#[$attr])*
+            #[test]
+            fn test_read_buffered_chunks() {
+                let data = $input;
+                let size = $len(&data);
+                let mut reader = BufReader::new(ReadStats::new(data));
+
+                let mut buffer = [0_u8; CHUNK_SIZE];
+                let mut chunks_read = 0;
+                while reader.read(&mut buffer[..]).expect("read must succeed") > 0 {
```

Comment:
> although it might be cumbersome to do so, maybe would be helpful if the error would indicate which read failed, by incorporating chunks_read? Unsure


same